design/20200326.extensible-certificate-controller.md

```diff
@@ -0,0 +1,388 @@
+    //     x509 data in the Secret, similar to what's done in earlier versions.
+    //     If there is a mismatch, an issuance is triggered.
+    //
+    // The final case above where no CertificateRequest resource exists is
+    // essential for backwards compatibility, as older CertificateRequest
+    // resources will not have a revision assigned so we cannot compare against
+    // them. In these cases, we fall back to the behaviour we use today of
+    // comparing the Certificate spec to the issued x509 certificate in the
+    // Secret.
+    //
+    // This condition may also be added by external API consumers to trigger
+    // a re-issuance manually for any other reason.
+    //
+    // It will be removed by the 'issuing' controller upon complete issuance.
+    CertificateConditionIssuing CertificateCondition = "Issuing"
+)
+```
+
+At the core of this proposal is the addition of the `status.iteration` field (an integer).
+This field indicates the current 'version' of the certificate.
+When a certificate is first created, the iteration is set to `1`.
+
+###
+#### Keymanager API fields
+
+We will add a new field to the `CertificateStatus` structure:
+
+```go
+package v1alpha3
+
+type CertificateStatus struct {
+    ...
+
+    // The name of the Secret resource 
+    // +optional
+    NextPrivateKeySecretName string `json:"nextPrivateKeySecretName,omitempty"`
+}
+```
+
+
+### Controller behaviour
+
+Implementing this proposal will require a complete replacement of the current
+`certificates` controller. Almost all areas of code will be replaced.
+
+The 'compare a `certificate.spec` with an x509 certificate' logic can and
+should be preserved to help maintain some semblance of backward compatibility
+for users upgrading from previous releases.
+
+#### keymanager
+
+The `keymanager` controller will be responsible for maintaining the
+`status.nextPrivateKeySecretName` field and any 'next private key' Secret
+resources that are owned by Certificates.
+
+* If the `Issuing` condition is `True`:
+    * If the `status.nextPrivateKeySecretName` field is not set:
+        * Check for existing 'next private key' Secret resources:
+            * If one matches, set the `status.nextPrivateKeySecretName` field.
+              This handles cache inconsistencies when we observe Secret
+              creation before updating `status.nextPrivateKeySecretName`.
+            * Otherwise, generate a new private key according to
+              `spec.privateKey` and store it in a new Secret resource.
+              Persist the name of the Secret as `status.nextPrivateKeySecretName`.
+    * If the `status.nextPrivateKeySecretName` field is set:
+        * If the Secret contains key data that matches `spec.privateKey`:
+            * do nothing
+        * If the Secret is not 'owned' by the Certificate, do nothing and log
+          an Event to inform the user.
+        * If the Secret is not labelled as a 'next private key', do nothing and
```

Comment:
Should this label be defined somewhere in the doc? Under API changes maybe?

Edit: I see this is mentioned below this section